rfcs/20240531-coo-sparse-encoding/README.md

```diff
@@ -0,0 +1,152 @@
+
+The addition of a new encoding will update the list for the sparse format kind: 
+
+```cpp
+/// Sparse encodings
+typedef enum {
+    /// Undefined sparse encoding kind, used for empty memory descriptors.
+    dnnl_sparse_encoding_undef = 0,
+    /// Compressed Sparse Row (CSR) encoding.
+    dnnl_csr,
+    /// An encoding that is used for an opaque storage schema for
+    /// tensors with unstructured sparsity. A memory descriptor with the
+    /// packed encoding cannot be used to create a memory object. It can
+    /// only be used to create a primitive descriptor to query the
+    /// actual memory descriptor (similar to the format tag `any`).
+    dnnl_packed,
+	/// Coordinate Sparse Encoding (COO)
+	dnnl_coo,
+} dnnl_sparse_encoding_t;
+```
+
+The memory object implementation in oneDNN supports underlying memory buffers that allows it to handle non-trivial data representations like sparse encodings.
+Based on this implementation, COO encoding will have a separate API for creating a memory descriptor which can specify its buffers. This can be done in three ways:
+
+### Option 1 - Sorted COO (Recommended)
+The memory descriptors will comprise of three buffers as defined for a sorted COO with one buffer each for data (index [0]), indices for the first dimension (index [1]) and indices for the second dimension (index [2]). The buffers for the data indices will share the datatype during declaration. 
+
+#### C API: 
+```cpp
+/// Creates a memory descriptor for COO encoding.
+///
+/// The created memory descriptor will describe a memory object that
+/// contains 3 buffers. The buffers have the following meaning and
+/// assigned numbers (index):
+///  - 0: values
+///  - 1: dim0_indices
+///  - 2: dim1_indices
+///
+/// @param memory_desc Output memory descriptor.
+/// @param ndims Number of dimensions.
+/// @param dims Array of dimensions.
+/// @param data_type Elements data type.
+/// @param nnz Number of non-zero entries.
+/// @param indices_dt Data type of indices.
+/// @returns #dnnl_success on success and a status describing the error
+///     otherwise.
+dnnl_status_t dnnl_memory_desc_create_with_coo_encoding(
+        dnnl_memory_desc_t *memory_desc, 
+		int ndims, 
+		const dnnl_dims_t dims,
+        dnnl_data_type_t data_type, 
+		dnnl_dim_t nnz, 
+		dnnl_data_type_t indices_dt);
+```
+
+#### C++ API:
+```cpp
+struct memory{
+	struct desc{
+		/// Function for creating a memory descriptor for COO sparse encodings.
+        ///
+		/// The created memory descriptor will describe a memory object that
+        /// contains 3 buffers. The buffers have the following meaning and
+        /// assigned numbers (index):
+        ///  - 0: values
+        ///  - 1: dim0_indices
+        ///  - 2: dim1_indices
+		///
+        /// @param adims Tensor dimensions.
+        /// @param adata_type Data precision/type.
+        /// @param nnz Number of non-zero entries.
+        /// @param index_dt Data type of indices.
```

Comment:
What sort of data types are expected for indices?

Reply:
Hi! Keeping the same data considerations as the CSR format, the datatype is expected to be s32 for the indices.